spec/bolt/module_installer_spec.rb

```diff
@@ -0,0 +1,95 @@
+require 'bolt/module_installer'
+
+describe Bolt::ModuleInstaller do
+  let(:puppetfile)     { @project + 'Puppetfile' }
+  let(:moduledir)      { @project + '.modules' }
+  let(:config)         { @project + 'bolt-project.yaml' }
+  let(:new_module)     { 'puppetlabs-pkcs7' }
+  let(:modules_config) { [{ 'name' => 'puppetlabs-yaml' }] }
+  let(:outputter)      { double('outputter', print_message: nil, print_puppetfile_result: nil) }
+  let(:pal)            { double('pal', generate_types: nil) }
+  let(:installer)      { described_class.new(outputter, pal) }
+
+  around(:each) do |example|
+    Dir.mktmpdir(nil, Dir.pwd) do |project|
+      @project = Pathname.new(project)
+      example.run
+    end
+  end
+
+  before(:each) do
+    conf = { 'modules' => [] }
+    File.write(config, conf.to_yaml)
+  end
+
+  context '#add' do
+    it 'returns early if the module is already declared' do
+      result = installer.add('puppetlabs-yaml', modules_config, puppetfile, moduledir, config)
+      expect(result).to eq(0)
+      expect(puppetfile.exist?).to eq(false)
+      expect(moduledir.exist?).to eq(false)
+    end
+
+    it 'errors if Puppetfile is not managed by Bolt' do
+      File.write(puppetfile, '')
+      expect { installer.add(new_module, modules_config, puppetfile, moduledir, config) }.to raise_error(Bolt::Error)
+    end
+
+    it 'updates files and installs modules' do
+      installer.add(new_module, modules_config, puppetfile, moduledir, config)
+
+      expect(puppetfile.exist?).to be(true)
+      expect(File.read(puppetfile)).to match(/mod "puppetlabs-pkcs7"/)
+
+      conf = YAML.safe_load(File.read(config))
+      expect(conf['modules']).to match_array([
+                                               { 'name' => 'puppetlabs-pkcs7' }
+                                             ])
+
+      expect(moduledir.exist?).to be(true)
+    end
+
+    it 'does not update version of installed modules' do
+      spec = 'mod "puppetlabs-yaml", "0.1.0"'
+      File.write(puppetfile, spec)
+      result = installer.add(new_module, modules_config, puppetfile, moduledir, config)
+
+      expect(result).to eq(0)
+      expect(File.read(puppetfile)).to match(/#{spec}/)
+    end
+
+    it 'updates version of installed modules if unable to resolve with pinned versions' do
+      spec = 'mod "puppetlabs-ruby_task_helper", "0.3.0"'
+      File.write(puppetfile, spec)
+      result = installer.add(new_module, [], puppetfile, moduledir, config)
+
+      expect(result).to eq(0)
+      expect(File.read(puppetfile)).not_to match(/#{spec}/)
+    end
+  end
+
+  context '#install' do
+    it 'errors if Puppetfile is not managed by Bolt' do
+      File.write(puppetfile, '')
+      expect { installer.install(modules_config, puppetfile, moduledir) }.to raise_error(Bolt::Error)
+    end
+
+    it 'installs modules forcibly' do
+      File.write(puppetfile, '')
+      expect { installer.install(modules_config, puppetfile, moduledir, force: true) }.not_to raise_error
```

Comment:
We should also check that it actually installs the modules, not just doesn't error.

Reply:
Added a check that the moduledir exists, but checking for specific modules causes problems in windows CI, as `Dir.children(moduledir)` doesn't seem to return anything. This seems more of an r10k / testing environment issue, as I've run into this exact problem before when moving our tests to github.

Reply:
We use `Dir.children` quite a bit in tests, i.e.: https://github.com/puppetlabs/bolt/blob/main/spec/bolt/project_migrate_spec.rb#L77 on both platforms. Is the issue that modules don't install to the right directory on Windows? Did you verify manually this works as expected on Windows?

Reply:
No, the issue is that the modules do not seem to install to the moduledir at all in Windows CI. Running the command on Windows correctly installs modules to the expected location.

It may make more sense to just have this test confirm that `#install_puppetfile` is called.